### ISSUE TITLE
Improve release process

### ISSUE THREAD

**Issue by frouioui** [vitessio/vitess#11861]
### Feature Description

Today, the release process is pretty straightforward. There is a pre-release step happening before the release day, a release day, and a post release step happening after the release day. During the release day we: generate the release notes, create a new release branche if needed, and run the release script to:
1. Transition from _"dev"_ mode to _"release"_ mode
2. Tag the release
3. Transition from _"release"_ to _"dev"_ mode

The "dev" mode is when the advertised version of our binaries says `vX.Y.Z-SNAPSHOT`, and _"release"_ mode when the version is simply `vX.Y.Z`. When we transition from _"release"_ to _"dev"_ mode, we increment the major or patch release increment and add the `-SNAPSHOT` label, this is the next version we want to release on the branch.

Once those three steps are executed, we need to push the changed to our release branch. The current approach is to have two distinct Pull Requests: one for the step 1 & 2, and another one for step 3.

There are a couple of issues with the current approach:
- There is no way to run the final commit we are releasing on through the CI. Even though the release is not changing bits of code that might break, we are not entirely safe from a human error while running the release process. Today, the script changes the codebase based on maintainers' input, create a commit and tags this commit. Having the same commit go through the CI once to confirm nothing is broken would increase safety but make the process longer as we would have to wait for CI to finish.
- The commit used to tag the release does not live on any branch. Since we use squash and merge to merge PRs in vitess, there is no way to find the original commit used for the release after we merge the first Pull Request.

The proposed solution is to divide the release day in two parts:
1. The first part where we freeze the release branch, generate the release notes and transition the codebase from _"dev"_ to _"release"_ mode. Once those actions are done, we would create a release Pull Request. We will then be able to run the future release commit through the CI and ensure it is green.
2. The second part is to do the actual release. Once the PR is merged, we would tag the release on the branch. Finally, we transition from _"release"_ to _"dev"_ mode on the branch and open a follow-up PR for this.

### Use Case(s)

n/a
